ml-agents/mlagents/trainers/ghost/trainer.py

```diff
@@ -0,0 +1,256 @@
+# # Unity ML-Agents Toolkit
+# ## ML-Agent Learning (Ghost Trainer)
+
+# import logging
+from typing import Dict, List, Any, cast
+
+import numpy as np
+import logging
+
+from mlagents.trainers.brain import BrainParameters
+from mlagents.trainers.policy import Policy
+from mlagents.trainers.tf_policy import TFPolicy
+
+from mlagents.trainers.trainer import Trainer
+from mlagents.trainers.trajectory import Trajectory
+from mlagents.trainers.agent_processor import AgentManagerQueue
+
+from mlagents.tf_utils.tf import TensorFlowVariables
+
+# logger = logging.getLogger("mlagents.trainers")
+
+LOGGER = logging.getLogger("mlagents.trainers")
+
+
+class GhostTrainer(Trainer):
+    def __init__(
+        self, trainer, brain_name, reward_buff_cap, trainer_parameters, training, run_id
+    ):
+        """
+        Responsible for collecting experiences and training trainer model via self_play.
+        :param trainer: The trainer of the policy/policies being trained with self_play
+        :param brain_name: The name of the brain associated with trainer config
+        :param reward_buff_cap: Max reward history to track in the reward buffer
+        :param trainer_parameters: The parameters for the trainer (dictionary).
+        :param training: Whether the trainer is set for training.
+        :param run_id: The identifier of the current run
+        """
+
+        super(GhostTrainer, self).__init__(
```

Comment:
Might be a bit late for this one - but I feel like we should rename "GhostTrainer" to "SelfPlayTrainer" and the config can have a category called `self_play`, which is more interpretable to the user than `ghost`